spec/models/product_spec.rb

```diff
@@ -0,0 +1,181 @@
+    end
+    let!(:comment3)do
+      FactoryBot.create :comment,
+                         product_id: product3.id,
+                         user_id: users[1].id
+    end
+
+    let!(:rate1) do
+      FactoryBot.create :rate,
+                         number_of_stars: 5,
+                         product_id: product1.id,
+                         user_id: users[0].id,
+                         comment_id: comment1.id
+    end
+    let!(:rate2) do
+      FactoryBot.create :rate,
+                         number_of_stars: 4,
+                         product_id: product2.id,
+                         user_id: users[1].id,
+                         comment_id: comment2.id
+    end
+    let!(:rate3) do
+      FactoryBot.create :rate,
+                         number_of_stars: 3,
+                         product_id: product3.id,
+                         user_id: users[1].id,
+                         comment_id: comment3.id
+    end
+
+    describe "filtering products" do
+      describe ".filter_by_product_size_id" do
+        it "has 2 products with first size" do
+          expect(
+            Product.filter_by_product_size_id(product_sizes[0].id).to_a
+          ).to eq([product1, product3])
+        end
+      end
+
+      describe ".filter_by_product_color_id" do
+        it "has 1 products with first size" do
+          expect(
+            Product.filter_by_product_color_id(product_colors[0].id).to_a
+          ).to eq([product1])
+        end
+      end
+
+      describe ".filter_by_max_cost" do
+        it "has 2 products with a cost less than 2000000" do
+          expect(
+            Product.filter_by_max_cost(2_000_000).to_a
+          ).to eq([product1, product2])
+        end
+      end
+
+      describe ".filter_by_min_cost" do
+        it "has 2 products with a cost greater than 2000000" do
+          expect(
+            Product.filter_by_min_cost(2_000_000).to_a
+          ).to eq([product2, product3])
+        end
+      end
+    end
+
+    describe "searching for products" do
+      describe ".search" do
+        it "has 1 products with keyword 'IPHONE'" do
+          expect(Product.search("IPHONE").to_a).to eq([product1])
+        end
+      end
+    end
+
+    describe "sort by products creation date" do
+      describe ".order_created_at" do
+        it "has 3 products sorted by creation date" do
+          expect(
+            Product.order_created_at.to_a
+          ).to eq([product3, product2, product1])
+        end
+      end
+    end
+
+    describe "sort top products" do
+      describe ".top_sellers" do
+        it "has 2 best selling products" do
+          expect(Product.top_sellers(2).to_a).to eq([product3, product2])
+        end
+      end
+
+      describe ".top_new" do
+        it "has 2 newest products" do
+          expect(Product.top_new(2).to_a).to eq([product3, product2])
+        end
+      end
+
+      describe ".top_rates" do
+        it "has 2 most voted products" do
```

Comment:
phân biệt describe, context và it

Reply:
- descrive và context đều để tạo group example, describe sử dụng với class, method, scores. context để sử dụng với các điền kiện như là "with" hay "when".
- it chứa example để mô tả một example ạ.